### PR TITLE
Add Gateway Intents

### PR DESCRIPTION
**Pull request by Lakelezz**
This adds support for gateway intents, you can specify these on the client via:

```rust
    let mut client = Client::new_with_extras(&token, |f| {
        f.intents(GatewayIntents::GUILDS).event_handler(Handler)
    })?;
```
By default, Serenity does not specify any intents.
Read more about gateway intents here: https://discordapp.com/developers/docs/topics/gateway#gateway-intents

Additionally, adds example 13 showcasing how to use gateway intents.

Closes: #815

### CODE REVIEW

File: src/client/bridge/gateway/intents.rs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+use serde::{
+    de::{Deserialize, Deserializer},
+    ser::{Serialize, Serializer},
+};
+
+#[derive(Copy, PartialEq, Eq, Clone, PartialOrd, Ord, Hash)]
+/// [Gateway Intents] will limit the events your bot will receive via the gateway.
+/// By default, no intents are specified by Serenity.
+///
+/// [Gateway Intents]: https://discordapp.com/developers/docs/topics/gateway#gateway-intents
```
Comment (arqunis) [inline src/client/bridge/gateway/intents.rs]:
```suggestion
/// [Gateway Intents] will limit the events your bot will receive via the gateway.
/// By default, no intents are specified by Serenity.
///
/// [Gateway Intents]: https://discordapp.com/developers/docs/topics/gateway#gateway-intents
#[derive(Copy, PartialEq, Eq, Clone, PartialOrd, Ord, Hash)]
```

File: src/gateway/shard.rs [hunk 2]
```diff
@@ -2,10 +2,11 @@ use crate::constants::{self, close_codes};
 use crate::internal::prelude::*;
 use crate::model::{
     event::{Event, GatewayEvent},
-    gateway::Activity,
+    gateway::{Activity},
```
Comment (arqunis) [inline src/gateway/shard.rs]: I'm assuming the addition of these braces was an accident?
```suggestion
    gateway::Activity,
```

File: src/client/bridge/gateway/intents.rs [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+pub struct GatewayIntents {
+    /// The flags composing gateway intents.
+    ///
+    /// # Note
+    /// Do not modify this yourself; use the provided methods.
```
Comment (arqunis) [inline src/client/bridge/gateway/intents.rs:16]: > use the provided methods

There aren't any, however, so this statement is misleading. You should add them in the same way we do for the [`Permissions` flags](https://github.com/serenity-rs/serenity/blob/current/src/model/permissions.rs#L272).

**Review by arqunis — CHANGES_REQUESTED** [review]

File: src/client/bridge/gateway/intents.rs (hunk 3 above)
Comment (Lakelezz) [reply src/client/bridge/gateway/intents.rs:16]: My bad, I forgot that, they will be added of course!

File: src/gateway/shard.rs (hunk 2 above)
Comment (Lakelezz) [reply src/gateway/shard.rs]: Yes, that was an accident.

**Review by arqunis — APPROVED** [review]

**Pull request merged**